Source/Core/Core/IOS/USB/Emulated/Skylander.cpp

```diff
@@ -0,0 +1,815 @@
+    return 0;
+  }
+  // If some data was requested from the Control Message, then the Interrupt message needs to
+  // respond with that data. Check if the queries queue is empty
+  if (!m_queries.empty())
+  {
+    result = m_queries.front();
+    m_queries.pop();
+    // This needs to happen after ~22 milliseconds
+    expected_time = 22000;
+  }
+  // If there is no relevant data to respond with, respond with the currentstatus of the Portal
+  else
+  {
+    result = g_skyportal.GetStatus();
+    expected_time = 2000;
+  }
+  expected_count = 32;
+  ScheduleTransfer(std::move(cmd), result, expected_count, expected_time);
+  return 0;
+}
+int SkylanderUSB::SubmitTransfer(std::unique_ptr<IsoMessage> cmd)
+{
+  DEBUG_LOG_FMT(IOS_USB, "[{:04x}:{:04x} {}] Isochronous: length={} endpoint={} num_packets={}",
+                m_vid, m_pid, m_active_interface, cmd->length, cmd->endpoint, cmd->num_packets);
+  return 0;
+}
+
+void SkylanderUSB::ScheduleTransfer(std::unique_ptr<TransferCommand> command,
+                                    std::array<u8, 64> data, s32 expected_count, u64 expected_time)
+{
+  command->FillBuffer(data.data(), expected_count);
+  command->ScheduleTransferCompletion(expected_count, expected_time);
+}
+
+void Skylander::Save()
+{
+  if (!sky_file)
+  {
+    return;
+  }
+
+  {
+    sky_file.Seek(0, File::SeekOrigin::Begin);
+    sky_file.WriteBytes(data.data(), 0x40 * 0x10);
+  }
+}
+
+void SkylanderPortal::Activate()
+{
+  std::lock_guard lock(sky_mutex);
+  if (m_activated)
+  {
+    // If the portal was already active no change is needed
+    return;
+  }
+
+  // If not we need to advertise change to all the figures present on the portal
+  for (auto& s : skylanders)
+  {
+    if (s.status & 1)
+    {
+      s.queued_status.push(3);
+      s.queued_status.push(1);
+    }
+  }
+
+  m_activated = true;
+}
+
+void SkylanderPortal::Deactivate()
+{
+  std::lock_guard lock(sky_mutex);
+
+  for (auto& s : skylanders)
+  {
+    // check if at the end of the updates there would be a figure on the portal
+    if (!s.queued_status.empty())
+    {
+      s.status = s.queued_status.back();
+      s.queued_status = std::queue<u8>();
+    }
+
+    s.status &= 1;
+  }
+
+  m_activated = false;
+}
+
+bool SkylanderPortal::IsActivated()
+{
+  std::lock_guard lock(sky_mutex);
+
+  return m_activated;
+}
+
+void SkylanderPortal::UpdateStatus()
+{
+  std::lock_guard lock(sky_mutex);
+
+  if (!m_status_updated)
+  {
+    for (auto& s : skylanders)
+    {
+      if (s.status & 1)
+      {
+        s.queued_status.push(0);
+        s.queued_status.push(3);
+        s.queued_status.push(1);
+      }
+    }
+    m_status_updated = true;
+  }
+}
+
+// Side:
+// 0x00 = right
+// 0x01 = left and right
+// 0x02 = left
+// 0x03 = trap
+void SkylanderPortal::SetLEDs(u8 side, u8 red, u8 green, u8 blue)
+{
+  std::lock_guard lock(sky_mutex);
+  if (side == 0x00)
+  {
+    this->m_color_right.red = red;
```

Comment:
`this->` is unneeded here and should be omitted (to be consistent with Dolphin's code style)